[RunAllTests] Fix #2944,  #2787, #298, #3007: Update interactions to support new localization structures

BenHenning opened the pull request:
## Explanation
Fix #2944
Fix #2787
Fix #298 (the prohibit bit isn't happening, but JSON loading will eventually be going away so that seems fine)
Fix #3007

This is the primary PR for bringing updated lesson structure support (mainly for explorations/questions) into the app. There were a few migrations in Oppia web since the last time we updated the app's structure (September), and most of them corresponded to the continued effort to bring answer localization support to the platform. At a high-level, the following structural changes affected the app (directly & indirectly):
- The answer & input types changed for item selection to correspond to a set of content IDs/a content ID (respectively) rather than actual options
- Customization args for item selection & multiple choice brought content IDs into them via SubtitledHtml
- DragAndDropSortInput's answer & input types changed for item selection to correspond to a list of set of content IDs rather than actual options
- TextInput's input parameters type has changed to be a list of strings, though I think currently creators can still only specify one
- TextInput's CaseSensitiveEquals rule was removed in favor of always using a case-insensitive check (i.e. the current 'Equals' rule). It seems all of TextInput's rules are now case-insensitive checks, and a few other providers needed to be updated in the codebase to match this.
- Worked examples in concept cards now have both explanations and questions, though the app is only parsing the explanation fields (and isn't yet using them--#352 is tracking adding support for questions and explanations long-term)

Handling all of the above core changes required substantially changing both structural & testing aspects of the codebase. Some additional changes were also needed due to asset & conversion limitations (I'll mention those in more detail below). For production behavior changes, most of the changes just needed to happen within the UI view models for interactions, and in the classifier provider classes in the domain layer. Nearly all other changes fall into the 'testing' category.

To ensure the app is behaving correctly, I updated our script that we're using to download the alpha assets to instead parse the old JSON files and convert them both to new JSON and textproto. However, I noticed that we were inconsistently handling different schema versions. Rather than trying to support multiple schemas versions, I removed the old lessons. This result in a lot of cleanups (e.g. some parsing code pathways could be cleaned up), but also required in a lot of test asset file changes (which are auto-generated from manual changes to the original assets) & a *lot* of test changes. To provide some context, I made the following changes to the assets to try and simplify things:
- test_exp_id_0, ratio_input_exploration, test_exp_id_1, test_exp_id_3 were all removed (which required rewriting much of StateFragmentTest and ExplorationProgressControllerTest, among others)
- test_exp_id_2 includes some additional answer groups to ensure all of RatioInput's rule types are being included (for tests)
- test_exp_id_2 was updated to include hints & a solution to migrate some existing domain tests that relied on a now-removed test exploration
- Manual fixes were needed in questions.json for past fields that were using camel case instead of snake case, and for missing fields/inconsistent structures. This led to some minor test fixes in QuestionAssessmentProgressControllerTest.

Regarding questions, while this does bring in updated questions for JSON and non-empty questions for proto (via a new QuestionListRecord proto), it does not actually bring proto question support yet (which is tracked by #2976). I spent about an hour trying to get this to work, and unfortunately the main issue is that the textproto->pb Bazel Starlark rule seems to not like the question proto since it's the only formatted proto that depends on another formatted proto (explorations). I didn't have more time to dig into it, though I suspect it's fixable.

Besides the above, other noteworthy changes here:
1. This fixes #3007 by enabling proto support in that test when run on Bazel. This has also been enabled for other tests in the PR chain to try and ensure different aspects of the proto pipeline are being covered in tests.
2. I fixed a lot of small style/line length/formatting/consistency issues throughout the codebase as I noticed them (particularly noteworthy is the StateRetriever class which had a bunch of areas for improvement). That unintentionally increases the length of the PR--sorry about that.

Note also that the changes seem to be passing on Espresso (or, at least, no new failures are observed): https://github.com/oppia/oppia-android/pull/2985#issuecomment-826030955.

BenHenning committed:
Add support for loading proto versions of lessons.

BenHenning committed:
Upgrade Espresso versions to work around build issues.

Note that I can't actually verify Espresso tests are working with these
versions since the Bazel project doesn't yet have Espresso tests set up.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Lint fix post-copying from #2927.

BenHenning committed:
Fix Glide in Bazel.

BenHenning committed:
Fix concept card name escaping.

Also, refactor HtmlParser to better support/handling different tag types
& prepare for LaTeX SVG image support.

BenHenning committed:
Copy over test fix from #2927.

BenHenning committed:
Add support for rendering LaTeX via SVGs.

Much of this is copied from #2927 so see that PR's commit history & the
PR corresponding to this commit for specific details.

BenHenning committed:
Ensure all thumbnails are opaque.

BenHenning committed:
Fix interactions to support new localization.

This fixes answer classification, submission, and viewing for drag &
drop, item selection, multiple choice, and text input interactions per
changes in the states structure to support lesson localization.

BenHenning committed:
Fix blurring for SVGs.

SVGs did not previous support blurring, & this change introduces a way
to leverage the existing bitmap blurrer to introduce blurring
functionality in a way that's consistent with bitmap thumbnails.

BenHenning committed:
Lint fixes post copying from #2927.

BenHenning committed:
Remove image loading annotation.

This will be added in a later branch in the chain.

BenHenning committed:
Add caching module to needed test suites.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Re-add image load annotation for this PR.

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

BenHenning committed:
Merge branch 'fix-locked-lesson-blurring' into add-support-for-item-selection-multiple-choice-drag-drop-localization

BenHenning commented:
/cc @rt4914 @vinitamurthi

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputEqualsRuleClassifierProvider.kt:
```diff
@@ -1,32 +1,35 @@
+    answer: SetOfTranslatableHtmlContentIds,
+    input: SetOfTranslatableHtmlContentIds
+  ): Boolean = answer.getContentIdSet() == input.getContentIdSet()
 }
```
More for my understanding, does this handle mismatched ordering as well?

BenHenning replied:
Yep since ``getContentIdSet()`` returns an actual ``Set``, and ``Set`` equality doesn't rely on ordering: https://docs.oracle.com/javase/8/docs/api/java/util/Set.html#equals-java.lang.Object-.

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/classify/rules/textinput/TextInputStartsWithRuleClassifierProvider.kt:
```diff
@@ -1,32 +1,36 @@
-    return answer.normalizeWhitespace().startsWith(input.normalizeWhitespace())
+  override fun matches(answer: String, input: TranslatableSetOfNormalizedString): Boolean {
+    val normalizedAnswer = answer.normalizeWhitespace()
+    return input.normalizedStringsList.any { normalizedAnswer.startsWith(it.normalizeWhitespace()) }
```
Does this handle case inequality?

BenHenning replied:
Nice catch--this is a gap I missed. Either we implemented this wrong originally, or as part of getting rid of CaseSensitiveEquals the other TextInput rules were updated to be case-insensitive. Fixed it in the latest batch of changes that should be pushed momentarily.

rt4914 commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/DragAndDropSortInteractionViewModel.kt:
```diff
@@ -28,18 +32,25 @@ class DragAndDropSortInteractionViewModel(
+  val choiceItems: List<DragDropInteractionContentViewModel> = _choiceItems
 
-  var isAnswerAvailable = ObservableField<Boolean>(false)
+  private var isAnswerAvailable = ObservableField<Boolean>(false)
```
Though this is not part of your PR but lets make it `val`

BenHenning replied:
Done. Changes should be available to review soon.

rt4914 reviewed (changes requested):
@BenHenning Are you planning to update json files in this PR as currently if it gets merged it will break the current develop.

BenHenning committed:
Add support for importing text protos as assets.

This actually provides Bazel-only support for converting text protos to
binary and including those as assets so that we can avoid checking in
binary protos to the codebase.

BenHenning committed:
Add gitignore for Android Studio Bazel plugin output.

BenHenning committed:
Import textproto versions of existing dev assets.

This imports local conversions of all existing dev assets as text proto
leveraging the Bazel-only conversion system to convert these to binary
protos and include them as assets.

This is a necessary step to both eventually replace JSON assets with
text proto, and to test that the new proto loading pipeline is working
as designed.

BenHenning committed:
Undo Espresso version change since it breaks Bazel tests.

BenHenning committed:
Fix tests.

Various attempts to fix existing tests + move some of the domain tests
over to actually using proto when running with Bazel.

App module tests are hanging currently for unknown reasons (only in
Bazel).

BenHenning committed:
Use correct Espresso core.

3.1.0 causes tests to hang indefinitely when run with Bazel.

BenHenning committed:
Move coroutine dispatchers.

This moves the coroutine dispatchers to a new threading subpackage. It
also simplifies the TestCoroutineDispatcher interface & implementations
(which required moving runUntilIdle to CoroutineExecutorServiceTest).

The test coroutine annotations were also moved to their own files.

BenHenning committed:
Refactor testing utilities build graph.

This introduces new Bazel libraries for:
- Test dispatchers
- Fake system/Oppia clocks
- Robolectric dependencies

BenHenning committed:
Introduce dedicated test for TestCoroutineDispatcher.

This also fixes some threading issues in the dispatcher, and clarifies
some of its API that was previously unclear (and can lead to subtle race
conditions).

BenHenning committed:
Add tests for TestCoroutineDispatcher.

This required a bunch of refactoring, adding the first examples of
dedicated Bazel test build files, and introducing a new test pattern for
sharing code between different implementations. Further, it refined the
API for TesCoroutineDispatcher & fixed some issues in the
implementations (especially Robolectric).

Tests verified as non-flaky in Robolectric across 1000 runs, and 0.2%
flaky across 1000 runs for Espresso (though that may be fixed; waiting
on a follow-up run to confirm stability).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Workaround ktlint semicolon issue.

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix test post-merge.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Move CoroutineDispatcher to be part of threading.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into fix-protobuf-duplication-issue

BenHenning committed:
Undo version upgrade.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	.gitignore
	app/src/sharedTest/java/org/oppia/android/app/story/StoryActivityTest.kt

BenHenning committed:
Fix broken tests.

BenHenning committed:
Lint fixes.

BenHenning committed:
Add a bit more proto loading test coverage.

BenHenning committed:
Comment fix.

BenHenning committed:
Lint fixes.

BenHenning committed:
Address earlier TODO.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	tools/BUILD.bazel

BenHenning committed:
Post-merge restructure.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	data/BUILD.bazel

BenHenning committed:
Post-merge fixes.

This removes unneeded dependencies from PersistentCacheStore, including
Glide.

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Add tests for new tag handlers.

Also, add tests for HtmlParser, and clean up some existing tests.

BenHenning committed:
Fix broken test build.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into fix-glide-image-loading

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Post-merge lint fixes.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

Conflicts:
	app/BUILD.bazel
	app/src/sharedTest/java/org/oppia/android/app/parser/HtmlParserTest.kt
	testing/BUILD.bazel
	utility/BUILD.bazel

BenHenning committed:
Lint fixes.

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Post-merge fix.

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

Conflicts:
	data/BUILD.bazel
	utility/src/main/java/org/oppia/android/util/caching/CacheAssetsLocally.kt

BenHenning committed:
Address pre-merge TODO.

BenHenning committed:
Disable local image loading.

Also, add specific LaTeX line in the first fractions interaction so that
the new SVG support can be tested.

BenHenning committed:
Fix typo.

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Docs + tests.

This adds more documentation & context for code added before without it.
It also adds a bunch of new testing coverage for changes. The tests are
particularly around the html part of the LaTeX loading pipeline, but not
the image parts.

More of both will be needed specifically for the image loading pipeline
changes.

BenHenning committed:
Docs + tests + refactor.

This adds documentation for all remaining components introduced in #2981,
adds some testing for UrlImageParser, and cleans up the image loading
pipeline to be a bit simpler & cleaner (though this will probably break
down-stream PRs relying on image transformations).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix lint failure & re-enable assets.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Clean up + tests.

Add tests where reasonable given current testing limitations and
developing time availability, and remove some dead code. Add new testing
utility (GenericViewMatchers) + corresponding Bazel package.

BenHenning committed:
Consolidate testing module view matchers.

This moves two other view matchers that are part of the testing module
into the new Espresso package.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

Conflicts:
	utility/src/main/java/org/oppia/android/util/parser/SvgPictureDrawable.kt

BenHenning committed:
Post-merge fixes + docs.

Add TODOs for tests that can't easily be added. Fill in missing
documentation. Fix post-merge issues to ensure SVGs can actually be
blurred (will be verified after this commit).

BenHenning committed:
Merge branch 'fix-locked-lesson-blurring' into add-support-for-item-selection-multiple-choice-drag-drop-localization

Conflicts:
	domain/src/main/java/org/oppia/android/domain/util/InteractionObjectExtensions.kt

BenHenning committed:
Add tests for structural changes.

Revamps StateFragmentTest to include a bunch of new interaction tests to
verify that the new structural changes are correct. These discovered
issues with the old asset converter that needed to be fixed, and new
textproto changes needed to be pulled in.

Removed a few now-unused (or should be unused) asset files.
ExplorationProgressControllerTest is now broken and has to be almost
entirely rewritten to support different explorations. The previous ones
couldn't reasonably be migrated without lots of manual work, so we
abandoned them.

BenHenning commented:
So a brief update on this: this PR is going to be *much* larger than I originally planned. The core structural changes that occurred in web that we're trying to bring in are just for multiple choice, item selection, drag & drop, and text input. Each of these were updated to support localized answers which requires a small structural change. However, these proliferate majorly throughout the codebase:
- Due to the size of the test lessons, I needed to create a script to migrate the old json format to the new one
- As part of the migration, a few issues came up that needed to be addressed (e.g. bugs in the migrator, and needs to add additional functionality to the existing test lessons)
- To simplify the migration, some old lessons were dropped in support (such as the welcome to Oppia lesson), partly due to the fact that the app was actually supporting multiple schema versions. This normalizes that.
- Removing some of the lessons required porting tests over to different lessons
- The new structural changes to answer types required pretty far reaching changes to existing tests
- The JSON loading pipeline & some of the surrounding support code need/benefited from upgrades

The main issue I ran into was questions. I thought we could get away with disabling questions for now (e.g. #3113), but I don't think that is going to work anymore: too many tests rely on questions functionality, and disabling so many regression tests for even a few months is likely to introduce a lot of issues when we try to re-enable it later. Because of this, I think I'm actually going to need to migrate our existing test questions even though the production version of the app (e.g. alpha MR2) is not going to support questions at the current time. This is going to take me a few more hours before I can get this PR into review.

BenHenning committed:
Rename test (& fix it).

Relevant fixes will be coming in a later commit. This is just committing
the file move.

BenHenning committed:
Fix/update json assets.

This includes fixes for existing content in textproto files, too, where
the previous conversion script had issues.

Assets that will no longer be included are removed (due to being too
hard to migrate).

The codebase will be broken at this point--thorough test changes are
needed, first.

BenHenning committed:
Fix a LOT of tests that break due to changes.

Fundamentally, this:
1. Migrates existing structural changes in tests for item selection,
text input, and drag and drop.
2. Introduces proto loading (when run on Bazel) for a few more tests,
including some UI tests.
3. Updates test suites to no longer depend on explorations that weren't
carried over in this change.
4. Cleans up a bunch of existing tests/line wrapping issues and other
things.

BenHenning committed:
Import structurally updates questions.

This includes proto questions even though they can't yet be imported.
Some tests needed adjustments due to fixes in the questions themselves.

BenHenning committed:
Merge branch 'develop' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'develop' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'develop' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'develop' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

BenHenning committed:
Merge branch 'develop' into fix-locked-lesson-blurring

BenHenning committed:
Merge branch 'fix-locked-lesson-blurring' into add-support-for-item-selection-multiple-choice-drag-drop-localization

BenHenning committed:
Post-merge lint fixes.

BenHenning committed:
Address reviewer comments.

Fix case sensitive checks for text input, and some minor cleanups in
drag & drop interaction view model.

BenHenning committed:
Fix broken Gradle test.

Another structural change: 'translation' is now the correct reference
for written translations rather than 'html'.

BenHenning commented:
Thanks for the initial review. FYI @rt4914 & @vinitamurthi that this is now ready for a full review.

@rt4914 I did mean to originally update the JSON structures when I sent this out, I just hadn't gotten to it yet. It turns out that led to a bunch of additional work I hadn't anticipated being as large as it ended up being. :)

Also,  please note that I haven't yet run the app tests via Espresso. I will do that tomorrow and fix any issues that arise. Please don't block on that detail unless you have specific concerns (in which case, please let me know).

vinitamurthi reviewed (approved):
LGTM!

rt4914 reviewed (approved):
LGTM (haven't run the espresso tests)

BenHenning commented:
Thanks for the reviews! It seems StateFragmentTest & ConceptCardFragmentTest are the only shared tests I changed in this PR, so I'm only testing those with Espresso. I'm using Robolectric passing as a rough proxy for other indirectly affected shared tests since running the whole suite is a bit too time consuming, and likely already has existing failures.

- Existing failures in StateFragmentTest (observed on develop): all loadImageRegion tests
- Observed failures in StateFragmentTest from this PR: none (all pass)
- Existing failures in ConceptCardFragmentTest (observed on develop):  all loadImageRegion tests
- Observed failures in ConceptCardFragmentTest from this PR: none (all pass)

It appears no new tests are failing for Espresso tests.

BenHenning commented:
Given no open comments, passing CI, and no new detected local Espresso failures, merging this.
